lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -213,26 +203,75 @@ let next dis s =
 
 let stop_on = [`Valid]
 
-(* find intersections, i.e. instructions which share bytes, but have different address,
-   and update destinations for those of them, which also share max address to
-   split a block in such place *)
+module Dest = struct
+  let rec compare_edge e e' =
+    if e = e' then 0
+    else
+      match e, e' with
+      | `Fall, _ -> -1
+      | `Jump, _ -> 1
+      | _ -> - (compare_edge e' e)
+
+  let compare (d,e) (d',e') =
+    match d, d' with
+    | None,None -> compare_edge e e'
+    | Some _, None -> 1
+    | None, Some _ -> -1
+    | Some d, Some d' ->
+       let x = Word.compare d d' in
+       if x = 0 then compare_edge e e'
+       else x
+
+  module Set = Set.Make(struct
+      type t = Brancher.dest [@@deriving sexp]
+      let compare = compare
+    end)
+
+end
+
+
+let equal_dest (d,e) (d',e') =
+  match d, d' with
+  | None, None -> e = e'
+  | Some _, None | None, Some _ -> false
+  | Some d, Some d' -> Word.(d = d') && e = e'
+
 let update_intersections disasm brancher s =
-  let next dis s = match s.roots with
-    | [] -> Dis.stop dis s
+  let next dis ((roots,dests) as state) = match roots with
```

Comment:
could please explain what is this function doing, and, most importantly, what are _intersections_ in your parlance?

Reply:
When I'm taking about intersections I mean instructions, which have intersected memory regions
And this function finds such instructions and updates the destinations table for those of them that have the same last byte and at least one common destination like it's described in details further. 
We save all the intersected instructions as a mapping `addr -> addr list`, where a key is the last address of instruction and data is a list of instructions addresses that share it. And if there is more then one instruction in a list then we need to find the destinations for each of them and update the destination table for those of the instructions which have at least one common destination.